### PR TITLE
Use C++17 "reflection" to deduce class / member / function names automatically

### PR DESCRIPTION
**Pull request by schaumb**
This simplifies the declarations, optionally.

### CODE REVIEW

File: include/nanobind/nanobind.h [hunk 1]
```diff
@@ -32,10 +32,12 @@
 #include <typeinfo>
 #include <utility>
 #include <new>
+#include <string_view>
```
[the remark this answers is not in the document]
Comment (schaumb) [reply include/nanobind/nanobind.h:35]: string_view used at compile-time string manipulation.

**Comment by schaumb** [comment]
It can recognize:
- class name 
  - template classes without template argument. So It can not able to recognize typedefs and usings.
    (`std::wstring` name is `basic_string`)
- member function names*
  - overloaded functions with `nb::overload_cast`
  - template functions with explicit template arguments `&fvec::emplace_back<float&&>`
- class static function names, with same techniques as member functions
- member variable names*
- class static variable names
- enum names
- enum values
- free function names, with same techniques as member functions
- free global lambda values as function
- bonus: free global values, with implementation of

```cpp
    template <auto glob_var>
    module_ &module_::def_global() {
      setattr(*this, detail::get_name<glob_var>(), cast(*glob_var, rv_policy::reference));
      return *this;
    }
    
    ...
    
 int global_counter;

    ...
    m.def_global<&global_counter>();
```



*VS cannot recognize:
- virtual functions name
- member variables name

File: include/nanobind/nanobind.h (hunk 1 above)
[the remark this answers is not in the document]
Comment (schaumb) [reply include/nanobind/nanobind.h]: I replaced it with a dummy implementation

**Comment by schaumb** [comment]
[Rerun](https://github.com/schaumb/nanobind/actions/runs/2036531879) tests without using string_view

**Comment by wjakob** [comment]
Dear @schaumb,

apologies for the delay, I finally had a moment to look at this PR. I see two major problems with it:

1. The implementation relies on non-standard/compiler-specific macros (`__PRETTY_FUNCTION__`, etc.). Just to be clear, nanobind already uses such extensions. For example, there are numerous `__attribute__(..)` peppered throughout the codebase. The big difference is that these only affect the optimality of the generated code. Things can still work when those extensions aren't available. That is not the case here.

2. This PR adds code (and likely, compilation cost) to the core parts of nanobind, to fix what I would consider a relatively minor problem. (I generally agree that having to write names twice can make writing binding code easier to write, but in the list of annoyances related to bindings, it does not make it into the top-10 list at least for myself.)

My suggestion would be that you find a way to realize this feature outside of the `nanobind` repository. For example, you could create an "automatic class" wrapper
```
#include <nanobind/nanobind.h>
template <typename T> aclass : public nanobind::class_<T> { .. }
```
that adds this auto-name-inteference logic.

**Comment by schaumb** [comment]
Thanks for the answer @wjakob .

I read the nanobind main page description, and I understand that the main goal is the simplicity of the codebase and keeping the compile-/runtime on the minimum.

I started this PR because for me (as one of the potential user) the interface matter too. 

This modification enables us to OPTIONALLY use the auto name deduction feature. So I disagree with your problems:

1. When this library is used on one or multiple major compilers, it's easier to use this simplified interface. These compiler-specific macros are broad-wide supported. So the following are remaining true after this PR:
> Supported compilers: Clang 7, GCC 8, MSVC2019 (or newer) are officially supported.

Based on the shortcomings of the MSVC compiler the interface can be designed to that every compiler works as same, without platform-specific `static_assert`.

2. The (force inlined) template member functions and default arguments have no compile-time/runtime cost when not used.
So when the compile-time and binary size matter, it can use the original, manually named definition bindings. When error prevention and detection (like copy/paste) is the main goal, this simplified interface can be used.

If the one `#include "nb_name.h"` line is a big overhead, it can be separate to an optionally includable header to not a 'core feature'.

I can accept that you have the decision about this library, but I think it's worth at least a discussion post when these concerns can be discussed and measured.

**Comment by wjakob** [comment]
This is a "nano" library, which means that the response for most "nice to have" features will be no to avoid severe feature-creep over time (which happened for pybind11). Changes like what you propose here can be perfectly realized outside of this nanobind using a simple wrapper, which means that they don't need to be part of the project. 

Note also that only Clang 7, GCC 8, MSVC2019  are officially supported, but the idea is very much that this project _should_ also work for other standards-compilant compilers. By depending on nonstandard extensions unfortunately make this a clear case where "no" can be the only answer to the PR.

**Pull request closed without merging**